python/tabby-eval/v2-modal/eval_code_completion.py

```diff
@@ -0,0 +1,164 @@
+    def modal_tabby_ready():
+        url = "{}/v1/health".format(endpoint)
+        headers = {
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+            "Authorization": f'Bearer {token}'
+        }
+
+        try:
+            response = httpx.get(url=url, headers=headers, timeout=5)
+            if response.status_code == 200:
+                logging.info("Server details: {}".format(response.json()))
+                return True
+            else:
+                return False
+        except Exception as e:
+            logging.error(f"Failed to check service health: {e}")
+            return False
+
+    while not modal_tabby_ready():
+        time.sleep(5)
+
+    logging.info("Modal tabby server ready!")
+
+
+def monitor_serve_output(process):
+    while True:
+        line = process.stdout.readline()
+        if not line:
+            break
+        logging.info(line.strip())
+
+
+def start_tabby_server(model):
+    logging.info("Starting tabby server for model {model}".format(model=model))
+
+    modal_env = os.environ.copy()
+    modal_env["MODEL_ID"] = model
+    modal_env["EMBEDDING_MODEL_ID"] = EMBEDDING_MODEL_ID
+
+    process = subprocess.Popen(args=["modal", "serve", "app.py"],
+                               stdout=subprocess.PIPE,
+                               stderr=subprocess.STDOUT,
+                               text=True,
+                               env=modal_env)
+
+    threading.Thread(target=monitor_serve_output, args=(process,)).start()
+
+    return process
+
+
+def send_sigint_to_process(process):
+    try:
+        os.kill(process.pid, signal.SIGINT)
+        logging.info("SIGINT signal sent successfully.")
+    except Exception as e:
+        logging.error(f"Failed to send SIGINT signal: {e}")
+
+
+def generate_predictions(endpoint, token, model, jsonl_file):
+    df = pd.read_json(jsonl_file, lines=True)
+    df_flat = json_normalize(df.to_dict(orient="records"))
+
+    headers = {
+        'Content-Type': 'application/json',
+        'Accept': 'application/json',
+        'Authorization': f'Bearer {token}'
+    }
+
+    predictions = []
+    for index, row in df_flat.iterrows():
+        payload = {
+            "language": row['language'],
+            "segments": {
+                "prefix": row['segments.prefix'],
+                "suffix": row['segments.suffix']
+            }
+        }
+
+        # TODO: Add retry logic, parallelism.
+        url = f"{endpoint}/v1/completions"
+        response = httpx.post(url=url, headers=headers, content=json.dumps(payload), timeout=10)
+
+        if response.status_code == 200:
+            predictions.append(response.json()['choices'][0]['text'])
+        else:
+            predictions.append("Request failed after retry.")
+
+    df_flat['prediction'] = predictions
+    formatted_timestamp = time.strftime("%Y%m%d%H%M%S")
+    prediction_jsonl_file = f"{formatted_timestamp}-{model.replace('/', '-')}.jsonl"
```

Comment:
instead of generate file name automatically, prefer passing output file path as a flag.

Reply:
use output_prediction_jsonl_file flag pass prediction file.